docs/doc/90-contributing/09-how-to-write-aggregate-functions.md

```diff
@@ -2,135 +2,126 @@
+    case_insensitive_combinator_desc: Vec<(String, CombinatorDescription)>,
 }
+impl AggregateFunctionFactory {
+  ...
+pub fn register(&mut self, name: &str, desc: AggregateFunctionDescription) {
+        let case_insensitive_desc = &mut self.case_insensitive_desc;
+        case_insensitive_desc.insert(name.to_lowercase(), desc);
+    }
+  ...
+}
+```
 
-### Understand the functions
-
-- The function `name` indicates the name of this function, such as `sum`, `min`.
-- The function `return_type` indicates the return type of the function, it may vary with different arguments, such as `sum(int8)` -> `int64`, `sum(uint8)` -> `uint64`, `sum(float64)` -> `float64`.
-- The function `nullable` indicates whether the `return_type` is nullable or not.
-
-Before we start to introduce the function `init_state`, let's ask a question first:
-
->  what's aggregate function state?
-
-It indicates the temporary results of an aggregate function. Because an aggregate function accumulates data in columns block by block and there will be some intermediate results after the aggregation. Therefore, the state must be mergeable, serializable.
-
-For example, in the `avg` aggregate function, we can represent the state like:
+Each registered function must implement both the `AggregateFunction` trait and the `AggregateFunctionFeatures` trait, where `AggregateFunctionFeatures` is more similar to `FunctionProperty` in `Scalar` in that they both store some properties of functions.
 
 ```rust
-struct AggregateAvgState<T: PrimitiveType> {
-    #[serde(bound(deserialize = "T: DeserializeOwned"))]
-    pub value: T,
-    pub count: u64,
+pub type AggregateFunctionRef = Arc<dyn AggregateFunction>;
+pub type AggregateFunctionCreator =
+    Box<dyn Fn(&str, Vec<Scalar>, Vec<DataType>) -> Result<AggregateFunctionRef> + Sync + Send>;
+pub struct AggregateFunctionDescription {
+    pub(crate) aggregate_function_creator: AggregateFunctionCreator,
+    pub(crate) features: AggregateFunctionFeatures,
 }
-
 ```
 
-- The function `init_state` initializes the aggregate function state, we ensure the memory is already allocated, and we just need to initial the state with the initial value.
-- The function `state_layout` indicates the memory layout of the state.
-- The function `accumulate` is used in aggregation with a single batch, which means the whole block can be aggregated in a single state, no other keys. The SQL query, which applies aggregation without group-by columns, will hit this function.
+## Function Composition
 
-Noted that the argument `_arrays` is the function arguments, we can safely get the array by index without index bound check because we must validate the argument numbers and types in function constructor.
+Unlike `Scalar`, which directly uses a Struct, `AggregateFunction` is a trait. Because aggregate functions accumulate data from the block and generate some intermediate results during the accumulation process.
```

Comment:
Unlike `Scalar` directly using a Struct, `AggregateFunction` is a trait because ...